src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php

```diff
@@ -297,13 +299,41 @@ public function viewAction(int $orderId, Request $request): Response
             'addOrderPaymentForm' => $addOrderPaymentForm->createView(),
             'changeOrderCurrencyForm' => $changeOrderCurrencyForm->createView(),
             'privateNoteForm' => $privateNoteForm->createView(),
-            'addProductToOrderForm' => $addProductToOrderForm->createView(),
-            'updateOrderProductForm' => $updateOrderProductForm->createView(),
             'updateOrderShippingForm' => $updateOrderShippingForm->createView(),
             'invoiceManagementIsEnabled' => $orderForViewing->isInvoiceManagementIsEnabled(),
         ]);
     }
 
+    /**
+     * @param int $orderId
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function addProductAction(int $orderId, Request $request): Response
+    {
+        $this->getCommandBus()->handle(AddProductToOrderCommand::withNewInvoice(
+            $orderId,
+            (int) $request->get('product_id'),
+            (int) $request->get('combination_id'),
+            (float) $request->get('price_tax_incl'),
+            (float) $request->get('price_tax_excl'),
+            (int) $request->get('quantity'),
+            false
+        ));
+
+        /** @var OrderForViewing $orderForViewing */
+        $orderForViewing = $this->getQueryBus()->handle(new GetOrderForViewing($orderId));
+
+        $products = $orderForViewing->getProducts()->getProducts();
+
+        return $this->render('@PrestaShop/Admin/Sell/Order/Order/Blocks/View/product.html.twig', [
+            'orderForViewing' => $orderForViewing,
+            'product' => end($products),
```

Comment:
Careful with `end` we had some bugs linked to it, if you can avoid it don't use it

Reply:
Replaced by `array_key_last`

Reply:
What do you mean by bugs with `end` ? and `array_key_last` will return the last key, not the last element :thinking:

Reply:
`end($products)` => `$products[array_key_last($products)]`